verilog/analysis/flow_tree.h

```diff
@@ -0,0 +1,133 @@
+// Copyright 2017-2022 The Verible Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef VERIBLE_VERILOG_FLOW_TREE_H_
+#define VERIBLE_VERILOG_FLOW_TREE_H_
+
+#include <bitset>
+#include <map>
+#include <string>
+#include <vector>
+
+#include "absl/status/status.h"
+#include "common/lexer/token_stream_adapter.h"
+#include "verilog/parser/verilog_token_enum.h"
+
+namespace verilog {
+
+// FlowTree class builds the control flow graph of a tokenized System-Verilog
+// source code. Furthermore, enabling doing the following queries on the graph:
+// 1) Generating all the possible variants (provided via a callback function).
+// 2) TODO(karimtera): uniquely identify a variant with a bitset.
+class FlowTree {
+ public:
+  using BitSet = std::bitset<128>;
+  using TokenSequenceConstIterator = verible::TokenSequence::const_iterator;
+
+  struct ConditionalBlock {
+    TokenSequenceConstIterator ifdef_iterator;
+    TokenSequenceConstIterator ifndef_iterator;
+    std::vector<TokenSequenceConstIterator> elsif_iterators;
+    TokenSequenceConstIterator else_iterator;
+    TokenSequenceConstIterator endif_iterator;
+  };
+
+  struct Variant {
+    // Contains the token sequence of the variant.
+    verible::TokenSequence sequence;
+
+    // The i-th bit in "macros_mask" is 1 when the macro (with ID = i) is
+    // assumed to be defined, o.w. it is assumed to be undefined.
+    BitSet macros_mask;
+
+    // The i-th bit in "assumed" is 1 when the macro (with ID = i) is visited or
+    // assumed (either defined or not), o.w. it is not visited (its value
+    // doesn't affect this variant).
+    //
+    // e.g.:
+    //  `ifdef A
+    //    `ifdef B
+    //    ...
+    //    `endif
+    //  `endif
+    //
+    // Consider the variant in which A is undefined,
+    // we notice that B doesn't affect the variant.
+    // Then the bit corresponding to B in "assumed" is 0.
+    BitSet assumed;
+  };
+
+  // Receive a complete token sequence of one variant.
+  // variant_sequence: the generated variant token sequence.
+  using VariantReceiver = std::function<bool(const Variant &variant)>;
+
+  explicit FlowTree(verible::TokenSequence source_sequence)
+      : source_sequence_(std::move(source_sequence)){};
+
+  // Generates all possible variants.
+  absl::Status GenerateVariants(const VariantReceiver &receiver);
+
```

Comment:
As an user of the API, I'd probably want to know what the meaning of each of the numbered macros is. So we should maybe have a
```
  const std::vector<TokenSequenceConstIterator> &GetUsedMacros();
```

That way, a user can associate bit N they get back in Variant in the BitSet with a meaningful name. Also they get an idea if the number of possible variants.